### ISSUE TITLE
RFC: Disable zero-dim CUDA tensors interacting with CUDA tensors on other devices

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#38205]
Today in PyTorch the following program is valid:

```
import torch

a = torch.rand([], requires_grad=True, device='cuda:0')
b = torch.rand(10, requires_grad=True, device='cuda:1')

c = a * b
```

Even though `a` and `b` are on different devices they can be multiplied since `a` is a zero-dim tensor. 

This seems to violate one of PyTorch's fundamental rules: no implicit device synchronization or data movement. As such, I propose we change our behavior to throw an error when a user attempts to do this, just like if they were trying to multiple non-zero dim CUDA tensors on different devices. 

As an aside, this does not work properly in current PyTorch either, see https://github.com/pytorch/pytorch/issues/36722. The device the kernel multiply kernel launches on is determined by the first CUDA argument, `a`, so a kernel is launched on device 0 to interact with `b`, on device 1, and a copy of `a` made by TensorIterator that's also placed on device 1. On systems without p2p access enabled this will even throw a runtime error.

**Comment by ngimel** [comment]
I think this is the right thing to do. As @mruberry notes, this behavior introduces implicit synchronization that we try to avoid, and due to current improper device guard setting it also introduces a significant performance regression on systems with p2p enabled, or crash and destruction of cuda context on the systems with p2p disabled. Rather than fixing this bug we should disallow undesirable behavior. 
cc @alband in case he already fixed autograd errors arising from allowing this behavior, these fixes probably won't be needed going forward.

**Comment by albanD** [comment]
The only change that happened is these lines: https://github.com/pytorch/pytorch/blob/1456515f1526560822ac5b2376af5d74e96a24fc/torch/csrc/autograd/engine.cpp#L616-L619
This is still needed when the origin is on the CPU, so we will need to keep it in.
